feat: data module alpha2: part 4

aaronc opened the pull request:
* implements `MsgServer` and `QueryServer` with test suites
* adds `ParseIRI`

clevinson commented on x/data/server/query_server.go:
```diff
@@ -10,55 +15,106 @@ import (
+			return nil, err
+		}
+
+		signerEntries = append(signerEntries, &data.SignerEntry{
```
Just thinking about this now – is there an attack vector in someone appending an arbitrary number of signatures to a given ContentHash, and this query getting bloated / slow? especially without pagination of signers...

technicallyty replied:
i agree here, i think it would be better to omit signers from the ByHash query and add a new query for a paginated list of signers on a certain content entry

clevinson commented on x/data/server/query_server.go:
```diff
@@ -10,55 +15,106 @@ import (
+			return err
+		}
+
+		entries = append(entries, entry)
```
So this looks like this query response will return a paginated list of entries, but each entry could have an unbounded number of signers inside the ContentEntry payload?  Maybe that's fine... but just flagging incase you see that as a potential problem down the line. Not sure what alternatives we have... maybe the `BySigner()` call shouldn't be expected to return ContentEntry's with full lists of all signers inside?

Just a thought.. ignore if you see this as a non-issue.

clevinson commented on x/data/server/testsuite/suite.go:
```diff
@@ -41,106 +42,168 @@ func (s *IntegrationTestSuite) TearDownSuite() {
+	s.Require().NoError(err)
+
+	// can query data and get timestamp
+	queryRes, err := s.queryClient.ByHash(s.ctx, &data.QueryByHashRequest{
```
Later do we expect to add a querier for `ByIRI`? Not for now, but perhaps in a follow-up. Seems like that would be useful, unless you don't expect app developers to be interacting with IRIs much themselves...

aaronc replied:
Let's handle this in a follow-up

clevinson commented on x/data/server/testsuite/suite.go:
```diff
@@ -41,106 +42,168 @@ func (s *IntegrationTestSuite) TearDownSuite() {
+	s.Require().NoError(err)
+	s.Require().NotNil(queryRes)
+	s.Require().NotNil(queryRes.Entry)
+	ts := queryRes.Entry.Timestamp
```
It would be nice in this test to actually verify the timestamp is accurate (especially as we are calling `AnchorData()` multiple times and asserting the second is a no-op.

clevinson commented on x/data/server/keys.go:
```diff
@@ -1,50 +1,56 @@
-	DataTablePrefix   byte = 0x3
+	IriIDTablePrefix      byte = 0x0
+	AnchorTimestampPrefix byte = 0x1
+	IDSignerPrefix        byte = 0x2
+	SignerIDPrefix        byte = 0x2
+	RawDataPrefix         byte = 0x3
```
Shouldn't `IDSignerPrefix` and `SignerIDPrefix` be distinct ?
```suggestion
	IDSignerPrefix        byte = 0x2
	SignerIDPrefix        byte = 0x3
	RawDataPrefix         byte = 0x4
```

clevinson reviewed (changes requested):
Nice! I'm seeing the pieces really come together here.

Some small non-blocking nits & clarifying q's, but there are also a few pieces in particular that i'd like to see responses on before approving:
- store key prefix bytes are not all distinct (i think this is a type-o)
- adding some clarification to why ParseIRI takes a generic `methodPrefix` when handling DID's, yet only seems to be able to return sdk.AccAddresses if the IRI is a DID
- i don't understand why `getIRIAndAnchor` always returns current blocktime, can you clarify if this is intentional, and if so why?

Otherwise, all looks good~!

robert-zaremba commented on x/data/server/testsuite/suite.go:
```diff
@@ -41,106 +42,168 @@ func (s *IntegrationTestSuite) TearDownSuite() {
+	}
+	s.Require().Contains(signers, s.addr1.String())
+	s.Require().Contains(signers, s.addr2.String())
+	s.Require().Empty(queryRes.Entry.Content)
```
why we don't declare 
```
require := s.Require()
```

at the beginning of the function - it would make the code easier to read.

robert-zaremba commented on x/data/iri.go:
```diff
@@ -111,3 +115,128 @@ var mediaTypeExtensions = map[MediaType]string{
+	}
+}
+
+func ParseIRI(methodPrefix string, iri string) (*ContentHash, sdk.AccAddress, error) {
```
I second that. Would be good to explain what's the expected iri, eg:
```
regen:<hashPart>.<ext>
```
What's `hashaPart` represents? It' seams it's more than a hash of something  - it's base58 with IRI version and a type. And we validate the last part of the hash: `digestAlg.Validate(hash)` - normally we validate hash against some other value. So, maybe we should rename it.

robert-zaremba reviewed (commented):
Left few comments, also Cory has few valid points re readability and 2 places with potential bugs.

aaronc commented:
Thanks for your reviews @robert-zaremba and @clevinson.

On a high-level I'd actually like some review of the IRI format we're using. Of all things related to this data module, we need to more or less look the IRI format in by mainnet launch because that will be used in documents and shouldn't change too much.

For instance, what do we even think about `regen:` as an IRI scheme? Should it be something more generic like a `hash:` IRI scheme? What about the way it's structured using extensions and specifically the `.rdf` extension for graph data?

Any thoughts @robert-zaremba @clevinson @blushi ?

blushi reviewed (commented):
> For instance, what do we even think about regen: as an IRI scheme? Should it be something more generic like a hash: IRI scheme? What about the way it's structured using extensions and specifically the .rdf extension for graph data?

Using `regen` as IRI scheme basically means that we reference something that leaves on regen ledger and per https://tools.ietf.org/html/rfc3986#section-3.1 refers to a specification for assigning identifiers within that scheme. So I guess it makes sense to use `regen:` because that clearly specifies restrictions on identifiers. What would be the use case for having a more generic IRI scheme? Or did you mean that we could have multiple hash-based schemes?

aaronc commented on x/data/server/query_server.go:
```diff
@@ -10,55 +15,106 @@ import (
+	}
+
+	store := ctx.KVStore(s.storeKey)
+	id := s.iriIDTable.GetOrCreateID(store, []byte(iri))
```
Oops! No we don't... we just want to get an existing ID or return `nil`... Good catch!

clevinson commented:
@aaronc I'm a fan of adding extensions to IRIs when they reference binary data, but now that i'm looking a bit more into examples of IRIs, i'm trying to think of how we actually expect this to be used. In looking at other [examples online](https://www.stardog.com/tutorials/data-model/#iris), I'm seeing a lot of use cases where the default namespace (e.g. `:Band` without a prefix, as it is referring to the local namespace) often consists of human readable text making the graph much easier to parse for end users.

Are there examples of how we expect to use human readable IRIs alongside hash based ones? I think it'd be nice to have some level of support for human readable IRIs in the default namespace. Perhaps we could provide support for this with some kind of name service for registering human readable IRIs? Alternatively (and probably the better option) would actually be to start with some base level schema that's defined in governance.

So if we want to support human readable IRIs, i'm wondering if we should differentiate btw prefixes of IRIs that correspond with `ContentHash`s and IRIs that are human-readable (potentially registered via a name service or managed by governance)?

Could we do something like `regen:hash:{CONTENT_HASH}.{EXT}` for all the ones we've previously discussed and maybe `regen:entity:{ENTITY_NAME}`? Is there best practicies for these? It seems like most examples I can find just use a raw "blank prefix" for this type of use case.

aaronc commented:
Fixed some of the most egregious errors here and removed `StoreRawData` (as per my concerns about GDPR and also unnecessarily bloating the ledger). There's still a lot of comments here for me to address, but at least I did find a bit of time to pick this up!

aaronc commented on x/data/server/msg_server.go:
```diff
@@ -11,156 +17,139 @@ import (
+
+var trueBz = []byte{1}
 
 func (s serverImpl) SignData(ctx types.Context, request *data.MsgSignDataRequest) (*data.MsgSignDataResponse, error) {
```
Let's maybe consider renaming to `Attest` in a follow-up

aaronc commented on x/data/server/msg_server.go:
```diff
@@ -2,164 +2,119 @@ package server
-//}
-
-//var emptyBz = []byte{0}
+type ToIRI interface {
```
it's not really a hash... let's leave as is for now and if we come up with something better we rename

aaronc commented on x/data/server/msg_server.go:
```diff
@@ -11,156 +17,139 @@ import (
-	//}
-	//
-	//return &data.MsgSignDataResponse{}, nil
+	iri, id, timestamp, err := s.getIRIAndAnchor(ctx, request.Hash)
```
we did some renaming to make it clearer and it looks like tests confirm it is a no-op

aaronc commented on x/data/server/query_server.go:
```diff
@@ -2,63 +2,120 @@ package server
+func (s serverImpl) getEntry(store sdk.KVStore, id []byte) (*data.ContentEntry, error) {
+	bz := store.Get(AnchorTimestampKey(id))
+	if len(bz) == 0 {
+		return nil, status.Error(codes.NotFound, "not found")
```
ID is mainly for internal purposes right now (not sent by user) so probably best not to leak it

clevinson commented on proto/regen/data/v1alpha2/query.proto:
```diff
@@ -62,10 +65,24 @@ message ContentEntry {
+  repeated string signers = 1;
+
+  // pagination is the pagination PageResponse.
+  cosmos.base.query.v1beta1.PageResponse pagination = 3;
```
```suggestion
  cosmos.base.query.v1beta1.PageResponse pagination = 2;
```

aaronc commented on x/data/iri.go:
```diff
@@ -111,3 +106,133 @@ var mediaTypeExtensions = map[MediaType]string{
+	const regenPrefix = "regen:"
+
+	if !strings.HasPrefix(iri, regenPrefix) {
+		return nil, sdkerrors.ErrInvalidRequest.Wrap(fmt.Sprintf("can't parse IRI %s without %s prefix", iri, regenPrefix))
```
@technicallyty we can simplify even further and just use wrapf. 

It would also be a good idea to add data module errors like InvalidIRI

ryanchristo reviewed (commented):
Looks good to me. A few questions/comments.

technicallyty commented on proto/regen/data/v1alpha2/query.proto:
```diff
@@ -20,6 +20,9 @@ service Query {
   }
+
+  // Signers queries by IRI.
+  rpc Signers(QuerySignersRequest) returns (QuerySignersResponse);
```
yep, that makes the most sense to me, changed here https://github.com/regen-network/regen-ledger/pull/236/commits/47ef05a9dd5d3beb471e82a15c50ba4d7f552e19

ryanchristo reviewed (approved):
It might be nice to add a few more test cases but otherwise looks good to me.

technicallyty commented on proto/regen/data/v1alpha2/query.proto:
```diff
@@ -12,8 +12,8 @@ option go_package = "github.com/regen-network/regen-ledger/x/data";
   // ByHash queries data based on its ContentHash.
-  rpc ByHash (QueryByHashRequest) returns (QueryByHashResponse) {
-    option (google.api.http).get = "/regen/data/v1alpha2/content/by_hash";
+  rpc ByIRI (QueryByIRIRequest) returns (QueryByIRIResponse) {
+    option (google.api.http).get = "/regen/data/v1alpha2/content/{iri}";
```
after a call with @clevinson we thought querying by iri makes more sense here. 

thoughts on this design? @aaronc @ryanchristo

ryanchristo replied:
Makes sense to me. More aligned with the signers query.

aaronc replied:
Sure but ideally we have both. Or maybe we add query methods to convert between IRIs and hashes but always query with IRIs by default?

aaronc commented:
@technicallyty @clevinson @ryanchristo what do we see remaining here to merge this? Are there some things that can be dealt with in follow-ups? Maybe we can break those out into issues if so...

ryanchristo commented:
> Are there some things that can be dealt with in follow-ups? Maybe we can break those out into issues if so...

I think we should merge and further discuss and address remaining items in follow-ups. Here are some of the items that I see as unresolved:

- [ ] add support for human readable IRIs
- [ ] use attestation terminology (e.g. rename `SignData` to `Attest`)
- [ ] add query methods to convert between IRIs and hashes but always query with IRIs by default
- [ ] improve test coverage (i.e. add more test cases for parsing IRI)

aaronc commented:
Thanks @ryanchristo. What do you mean by " add support for human readable IRIs"?

ryanchristo commented:
> What do you mean by "add support for human readable IRIs"?

This was something @clevinson brought up in https://github.com/regen-network/regen-ledger/pull/236#issuecomment-776411378. If we want to continue exploring this idea, we should open an issue or discussion to do so but I don't see it as a priority.

Also, I would be happy to open issues for the items listed following a second approval.

aaronc commented:
Sure please feel free to open issues 👍

clevinson reviewed (approved):
utACK.

Thanks @technicallyty for wrapping this up!
